### PR TITLE
Add org.jetbrains.annotations support

### PR DESCRIPTION
**Pull request by FirokOtaku**
I'm trying to add [`org.jetbrains.annotations`](https://github.com/JetBrains/java-annotations) support to core API of this lib. This will give better code completion in IntelliJ IDEA environment and should have no side effects.

Since I do not fully understand the code of this lib, the changes may contain errors and you may need to make a code review.

### CODE REVIEW

File: src/main/java/jssc/SerialPort.java [hunk 1]
```diff
@@ -241,7 +243,46 @@ else if(portHandle == SerialNativeInterface.ERR_INCORRECT_SERIAL_PORT){
      *
      * @throws SerialPortException if exception occurred
      */
-    public boolean setParams(int baudRate, int dataBits, int stopBits, int parity) throws SerialPortException {
+    public boolean setParams(
```
Comment (tresf) [inline src/main/java/jssc/SerialPort.java:246]: These are NOT magic values.

**Comment by tresf** [comment]
@FirokOtaku thanks for the enthusiasm.  I'm unsure what type of benefit this would provide to the codebase, so perhaps you can share a quick video explaining how this helps.  From a high-level, forcing contributors and maintainers to decorate code with more commenting seems a bit of overkill given how simple and straightforward the codebase is, but perhaps there's a use-case that's unobvious from the PR.

This project "JSSC" was originally written to be a Java Simple Serial Connector and the original author's intentions were to make the library as small and simple as possible.  Although these annotations don't change the size of the library, they do change the simplicity of it, so I think its use should be warranted.

**Comment by FirokOtaku** [comment]
> I'm unsure what type of benefit this would provide to the codebase, so perhaps you can share a quick video explaining how this helps.

Sure. Here is a short example:

```java
package firok.test.annotation;

import org.intellij.lang.annotations.MagicConstant;

public class Main
{
    public static final int MAGIC_VALUE_1 = 1, MAGIC_VALUE_2 = 2, MAGIC_VALUE_3 = 2;
    public static void funWithoutAnnotation(int value) 
    {
        switch (value)
        {
            case 1, 2, 3 -> { /* okay */ }
            default -> throw new IllegalArgumentException();
        }
    }
    public static void funWithAnnotation(
            @MagicConstant(intValues = {
                    MAGIC_VALUE_1,
                    MAGIC_VALUE_2,
                    MAGIC_VALUE_3,
            })
            int value
    ) {
        switch (value)
        {
            case 1, 2, 3 -> { /* okay */ } // with warning
            default -> throw new IllegalArgumentException();
        }
    }
    public static void main(String[] args)
    {
        // without @MagicConstant annotation,
        // IDEA will not give any warning even users are already using wrong values
        funWithoutAnnotation(1); // acceptable values, but no
        funWithoutAnnotation(2);
        funWithoutAnnotation(3);
        funWithoutAnnotation(4); // wrong value, but no warning

        funWithAnnotation(1); // acceptable values, IDEA will still give warning
        funWithAnnotation(2);
        funWithAnnotation(3);
        funWithAnnotation(4); // wrong value, with warning
        funWithAnnotation(MAGIC_VALUE_1); // acceptable values with no warning
        funWithAnnotation(MAGIC_VALUE_2);
        funWithAnnotation(MAGIC_VALUE_3);
    }
}
```

Screenshots:

all

On the users' side, IDEA will warn the user that they are not using the APIs in a proper way or even using a wrong value.

users' side

On the developers' side, when developer using those parameters, IDEA will give warning if the flow control code is using magic values rather than constants.

developers' side

After replacing magic values by constants, warnings disappear.

after

----

> From a high-level, forcing contributors and maintainers to decorate code with more commenting seems a bit of overkill given how simple and straightforward the codebase is, but perhaps there's a use-case that's unobvious from the PR.

Yes, this do need maintainers pay more time and energy. But this could help users and maintainers themselves discovering bugs at coding, rather than at runtime (when they are coding in IDEA or other Jetbrains IDEs), and can help maintaining better coding style.

And yes, this is something like platform-specific lib (or we should call it a IDE-specific lib here) and many one do not like this kind of libs. If you feel uncomfortable to it, just don't mind and close the PR.

----

> These are NOT magic values.

Umm okay. I'm new to serial port communication and do not clearly understand those concepts. I saw you mentioned that those APIs should take values from `XXX_`-prefixed constants and think they may be something like magic value.

----

> This is better refactored into an Enum or something that doesn't use magic values.

Yes, that's better.

----

BTW, `org.jetbrains.annotations` has lots of other useful annotations too. For example:

* `@MagicConstant#flags`/`@MagicConstatn#flagsFromClass` can be used for bitwise operation magic value: `fun(MAGIC_VALUE_1 | MAGIC_VALUE_2)`
* `@NotNull`/`@Nullable` can be marked on parameters or function return values
  * 
  * 
  * 
* `@Language` can be used on string values and give code highlights to string contents
  *

**Comment by tresf** [comment]
> Umm okay. I'm new to serial port communication and do not clearly understand those concepts. I saw you mentioned that those APIs should take values from XXX_-prefixed constants and think they may be something like magic value.

Right, and this is where I believe this PR is a bit overkill.

* This library is for communicating with a serial port
* Serial ports have industry-known values, such as `9600` baud
* The underlying APIs generally prefer to take these values as-is.
* `BAUDRATE_9600` is `9600`, so it's a [stretch to call it a "Magic Number"](https://en.wikipedia.org/wiki/Magic_number_(programming)).  If someone wants to submit a number value directly, they should be able to.

... on the other hand, `OS_LINUX` has no good modern-day justification to be a number.  It would be better refactored into an Enum.

> ```
> switch (value) [...]
> ```

Since most of the numbers you're referring to are passed to native APIs, I don't think using a switch case is a valid justification.

> @MagicConstant#flags/@MagicConstatn#flagsFromClass can be used for bitwise operation magic value: fun(MAGIC_VALUE_1 | MAGIC_VALUE_2)

In my experience with libraries, bitwise variables normally use a form of integer notation (e.g. Hex) which make them glaringly obvious to the developer that they can be combined.  If a developer does not understand this, it's not the job of one-and-only-one particular IDE to train them.

> But this could help users and maintainers themselves discovering bugs at coding, rather than at runtime

I do not see a valid example of a runtime bug as it pertains to this project's codebase.  This would be a step in the right directly as to providing efficacy as to this proposal.

> @Language can be used on string values and give code highlights to string contents

I do not see a valid example of a string highlight candidate in this codebase.  This would be a step in the right directly as to providing efficacy as to this proposal.

I believe what this PR misses is that most of the constants in this codebase:
1. Are not magic
2. Are passed as-is to the native components of the API
3. If they are magic, should be refactored into Enums

> If you feel uncomfortable to it, just don't mind and close the PR.

I believe -- in the case of this project -- the decorators are not helpful because they do not solve a particular problem.   I believe that the decorators are important for projects which use magic numbers more loosely to represent arbitrary concepts (such as our OS detection), which is often an anti-pattern that should be fixed with code, not documentation.  It seems to be older APIs (such as this one) which use constants over Enums, likely due to compatibility with JDK < 1.5.

I would be much happier with annotations that have a real-world justification as they pertain to this project, not theoretical.   I hope that's OK.  Please also accept my gratitude for taking the time to explain how these annotations work with the IDE, it's greatly appreciated.

**Comment by FirokOtaku** [comment]
@tresf Sorry for the late reply.

> In my experience with libraries, bitwise variables normally use a form of integer notation (e.g. Hex) which make them glaringly obvious to the developer that they can be combined. If a developer does not understand this, it's not the job of one-and-only-one particular IDE to train them.

Yes, if one guy does even not know how to use bitwise-flags, it is not the responsibility of any IDE or lib devs to train them. But the point of `@MagicConstant` is not "train", it is more like an interactive comment. For any users are using wrong values or bad code styles, IDE can give a warning and urge them using a right one.

> I do not see a valid example of a string highlight candidate in this codebase. This would be a step in the right direction as to providing efficacy as to this proposal.

Yes, the reason I mentioned that is not to say that can be used in current project, but just to mention that there are some other useful tools can be used in `org.jetbrains.annotations` lib which might be used in other projects.

> Right, and this is where I believe this PR is a bit overkill.
>
> If someone wants to submit a number value directly, they should be able to.

Thanks for the explaination. I understand your opinion and will close the PR then. Sorry for bothering you so much.

**Comment by tresf** [comment]
No bother at all.  Thank you for the recommendation, even if it's not accepted, it's greatly appreciated.

**Pull request closed without merging**